examples/localvideofilter/src/helpers.js

```diff
@@ -25,20 +31,34 @@ function displayLocalVideo(video) {
  */
 function filterLocalVideo(video, filtered, name) {
   var canvas = document.createElement('canvas');
+  var mediaStreamTrack = video.srcObject.getVideoTracks()[0];
+  var settings = mediaStreamTrack.getSettings();
+  canvas.width = settings.width;
+  canvas.height = settings.height;
+
   var context = canvas.getContext('2d');
-  canvas.width = VIDEO_WIDTH;
-  canvas.height = VIDEO_HEIGHT;
+  var filterValue = name === 'blur' ? '5px' : '100%';
+  context.filter = name === 'none' ? '' : name + '(' + filterValue + ')';
+  clearTimeout(filterTimeout);
+
+  var isSafari = /Safari/.test(navigator.userAgent)
+    && !/Chrome/.test(navigator.userAgent);
 
   function filterVideoFrame() {
-    context.drawImage(video, 0, 0, canvas.width, canvas.height);
-    var imageData = context.getImageData(0, 0, canvas.width, canvas.height);
-    context.putImageData(filters[name](imageData), 0, 0);
-    requestAnimationFrame(filterVideoFrame);
+    var renderStartTime = Date.now();
+    context.drawImage(video, 0, 0);
+    if (isSafari) {
+      var imageData = context.getImageData(0, 0, canvas.width, canvas.height);
+      context.putImageData(filters[name](imageData), 0, 0);
+    }
+    var renderDelay = Date.now() - renderStartTime;
+    var interFrameDelay = Math.round(1000 / settings.frameRate);
+    filterTimeout = setTimeout(filterVideoFrame, interFrameDelay - renderDelay);
   }
 
-  var stream = canvas.captureStream(30);
+  var stream = canvas.captureStream(settings.frameRate);
```

Comment:
Hardcoding the initial frame rate is not efficient because this remains constant if the track changes its frame rate. For example, if the initial frame rate of the track is very slow, you'll end up a very slow processed track although the original track has already increased frame rate.

What you can do here is to set this to zero, then inside `filterVideoFrame` method right after you draw the processed image, you can call `processedTrack. requestFrame()`. This makes sure the input and output FPS are always in sync. See our implementation for reference. https://github.com/twilio/twilio-video.js/blob/master/lib/media/track/videotrack.js#L150

Reply:
I'm afraid [requestFrame](https://developer.mozilla.org/en-US/docs/Web/API/CanvasCaptureMediaStreamTrack/requestFrame) is not supported in Firefox or Safari, so we cannot use it.

Reply:
On further inspection it is available in Safari, but not in Firefox. However, I don't think we have to worry about varying frame rates on the LocalVideoTrack in this example, since we don't use restart.